ratelimits/limiter.go

```diff
@@ -0,0 +1,214 @@
+package ratelimits
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/jmhodges/clock"
+)
+
+// ErrInvalidCost indicates that the cost specified was <= 0.
+var ErrInvalidCost = fmt.Errorf("invalid cost, must be > 0")
+
+// ErrInvalidCostForCheck indicates that the check cost specified was < 0.
+var ErrInvalidCostForCheck = fmt.Errorf("invalid check cost, must be >= 0")
+
+// ErrInvalidCostOverLimit indicates that the cost specified was > limit.Burst.
+var ErrInvalidCostOverLimit = fmt.Errorf("invalid cost, must be <= limit.Burst")
+
+var ErrBucketAlreadyFull = fmt.Errorf("bucket already full")
+
+type Limiter struct {
+	// defaults stores default limits by 'name'.
+	defaults limits
+
+	// overrides stores override limits by 'name:id'.
+	overrides limits
+	source    source
+	clk       clock.Clock
+}
+
+func NewLimiter(clk clock.Clock, source source, limitsPath, overridesPath string) (*Limiter, error) {
+	limiter := &Limiter{source: source, clk: clk}
+
+	defaults, err := loadLimits(limitsPath)
+	if err != nil {
+		return nil, err
+	}
+	limiter.defaults = defaults
+
+	if overridesPath == "" {
+		// No overrides specified.
+		limiter.overrides = make(limits)
+		return limiter, nil
+	}
+
+	overrides, err := loadLimits(overridesPath)
+	if err != nil {
+		return nil, err
+	}
+	limiter.overrides = overrides
+
+	return limiter, nil
+}
+
+type Decision struct {
+	// Allowed is true if the bucket has the capacity to allow the request.
+	Allowed bool
+
+	// Remaining is the number of requests remaining in the bucket.
+	Remaining int
+
+	// RetryIn is the duration the client must wait before they're allowed to
+	// make a request.
+	RetryIn time.Duration
+
+	// ResetIn is the duration the client would need to wait before the bucket
+	// reaches it's maximum capacity.
+	ResetIn time.Duration
+
+	// newTAT is the Theoretical Arrival Time of the next possible request.
+	newTAT time.Time
+}
+
+// Check returns a *Decision indicating whether capacity exists to service the
+// cost of the request for the provided limit Name and client id, but does not
+// spend the cost of the request. Most callers should use Spend instead. The
+// *Decision returned will always include the number of requests remaining in
+// the bucket, the duration the client must wait before they're allowed to make
+// another request, and the duration the client should wait before the bucket
```

Comment:
the number of requests remaining in the bucket _after_ the checked cost is spent? This doc should be clearer whether the returned Decision is a representation of what the bucket looks like now or what it _would_ look like if the caller actually called Spend.

Reply:
Thanks I'll rewrite this comment and the one for Spend to provide the correct emphasis.